state/indexer/sink/psql/psql.go

```diff
@@ -55,116 +56,182 @@ func (es *EventSink) DB() *sql.DB { return es.store }
-
-	sqlStmtEvents := sq.
-		Insert(TableEventTx).
-		Columns("key", "value", "height", "hash", "tx_result_id", "created_at", "chain_id").
-		PlaceholderFormat(sq.Dollar).
-		Suffix("ON CONFLICT (key,hash)").
-		Suffix("DO NOTHING")
-
-	ts := time.Now()
-	for _, tx := range txr {
-		txBz, err := proto.Marshal(tx)
-		if err != nil {
-			return err
-		}
+// insertEvents inserts a slice of events and any indexed attributes of those
+// events into the database associated with tx.
+//
+// If txID > 0, the event is attributed to the transaction with that ID;
+// otherwise it is recorded as a block event.
+func insertEvents(tx *sql.Tx, blockID, txID uint32, evts []abci.Event) error {
+	// Populate the transaction ID field iff one is defined (> 0).
+	var txIDArg interface{}
+	if txID > 0 {
+		txIDArg = txID
+	}
 
-		sqlStmtTxResult = sqlStmtTxResult.Values(txBz, ts)
+	// Add each event to the events table, and retrieve its row ID to use when
+	// adding any attributes the event provides.
+	for _, evt := range evts {
+		// Skip events with an empty type.
+		if evt.Type == "" {
+			continue
+		}
 
-		// execute sqlStmtTxResult db query and retrieve the txid
-		r, err := sqlStmtTxResult.Query()
+		eid, err := queryWithID(tx, `
+INSERT INTO `+tableEvents+` (block_id, tx_id, type) VALUES ($1, $2, $3)
+  RETURNING rowid;
+`, blockID, txIDArg, evt.Type)
 		if err != nil {
 			return err
 		}
-		defer r.Close()
 
-		if !r.Next() {
-			return nil
+		// Add any attributes flagged for indexing.
+		for _, attr := range evt.Attributes {
+			if !attr.Index {
+				continue
+			}
+			compositeKey := evt.Type + "." + attr.Key
+			if _, err := tx.Exec(`
+INSERT INTO `+tableAttributes+` (event_id, key, composite_key, value)
+  VALUES ($1, $2, $3, $4);
+`, eid, attr.Key, compositeKey, attr.Value); err != nil {
+				return err
+			}
 		}
+	}
+	return nil
+}
 
-		if err := r.Scan(&txid); err != nil {
-			return err
+// makeIndexedEvent constructs an event from the specified composite key and
+// value. If the key has the form "type.name", the event will have a single
+// attribute with that name and the value; otherwise the event will have only
+// a type and no attributes.
+func makeIndexedEvent(compositeKey, value string) abci.Event {
+	i := strings.Index(compositeKey, ".")
+	if i < 0 {
+		return abci.Event{Type: compositeKey}
+	}
+	return abci.Event{Type: compositeKey[:i], Attributes: []abci.EventAttribute{
+		{Key: compositeKey[i+1:], Value: value, Index: true},
+	}}
+}
+
+// IndexBlockEvents indexes the specified block header, part of the
+// indexer.EventSink interface.
+func (es *EventSink) IndexBlockEvents(h types.EventDataNewBlockHeader) error {
+	ts := time.Now().UTC()
+
+	return runInTransaction(es.store, func(tx *sql.Tx) error {
+		// Add the block to the blocks table and report back its row ID for use
+		// in indexing the events for the block.
+		blockID, err := queryWithID(tx, `
+INSERT INTO `+tableBlocks+` (height, chain_id, created_at)
```

Comment:
I considered that, but it would be a change of behaviour: the previous implementation attributes the same creation time to all the events for a block or a transaction. If we have the database populate the field, it may give different values for different events in the same object (block/tx).

Maybe that doesn't matter? In which case I'd be happy to do that. But I did it explicitly to preserve the way the previous implementation did it.

Reply:
I think keeping consistent behavior is a very valid reason to not make the change I suggested.